### ISSUE TITLE
`unifyfs terminate` utility fails after running an example

### ISSUE THREAD

**Issue by CamStan** [LLNL/UnifyFS#519]
### Describe the problem you're observing
The `unifyfs terminate` utility doesn't actually stop the servers **after** having run one or more of the example programs.

### Describe how to reproduce the problem

1. Start the server(s):
``` bash
#!/bin/sh

export UNIFYFS_DAEMONIZE=off
export UNIFYFS_LOG_VERBOSITY=5
export UNIFYFS_LOG_DIR=/path/to/log/dir

export UNIFYFS_LOGIO_SPILL_DIR=/dev/shm
export UNIFYFS_LOGIO_SPILL_SIZE=$((2 * 2 ** 30))

export UNIFYFS_SHAREDFS_DIR=/path/to/shared/dir

/path/to/UnifyFS/install/bin/unifyfs start -d -S $UNIFYFS_SHAREDFS_DIR -m /unifyfs -e /path/to/UnifyFS/install/bin/unifyfsd

sleep 10

echo "pid(s) of unifyfsd:"
jsrun -r1 pidof unifyfsd
```
which prints out the debug information along with the pid of `unifyfsd`:
```
⋮
pid(s) of unifyfsd:
22671
```

2. Run a `write` or `writeread` example:
``` bash
jsrun -n1 -r1 /path/to/UnifyFS/install/libexec/write-gotcha -v 2> err.out
```

3. Stop the server:
``` bash
#!/bin/sh

/path/to/UnifyFS/install/bin/unifyfs terminate -d

sleep 10

echo "pid(s) of unifyfsd:"
jsrun -r1 pidof unifyfsd
```
Which results in the same pid still being printed:
```
⋮
pid(s) of unifyfsd:
22671
```
-----
However, when I simply start the server and then stop it without running an example in between, it works fine.

### Include any warning or errors or releveant debugging data
The end of the server log after running an example followed by `unifyfs terminate` doesn't appear to complete:
```
2020-06-15T17:32:35 tid=22671 @ exit_request() [server/src/unifyfs_server.c:153] exit requested
2020-06-15T17:32:35 tid=22671 @ main() [server/src/unifyfs_server.c:398] starting service shutdown
2020-06-15T17:32:35 tid=22671 @ main() [server/src/unifyfs_server.c:403] stopping service manager thread
2020-06-15T17:32:35 tid=22680 @ sm_service_reads() [server/src/unifyfs_service_manager.c:414] service manager thread exiting
2020-06-15T17:32:35 tid=22671 @ svcmgr_fini() [server/src/unifyfs_service_manager.c:297] locking service manager state
2020-06-15T17:32:35 tid=22671 @ svcmgr_fini() [server/src/unifyfs_service_manager.c:303] unlocking service manager state
2020-06-15T17:32:35 tid=22671 @ main() [server/src/unifyfs_server.c:406] cleaning run state
2020-06-15T17:32:35 tid=22671 @ cleanup_application() [server/src/unifyfs_server.c:682] cleaning application 1511587981
```
As the server log from simply starting and stopping the server has additional steps at the end:
```
2020-06-15T17:36:44 tid=30932 @ exit_request() [server/src/unifyfs_server.c:153] exit requested
2020-06-15T17:36:44 tid=30932 @ main() [server/src/unifyfs_server.c:398] starting service shutdown
2020-06-15T17:36:44 tid=30932 @ main() [server/src/unifyfs_server.c:403] stopping service manager thread
2020-06-15T17:36:44 tid=30940 @ sm_service_reads() [server/src/unifyfs_service_manager.c:414] service manager thread exiting
2020-06-15T17:36:44 tid=30932 @ svcmgr_fini() [server/src/unifyfs_service_manager.c:297] locking service manager state
2020-06-15T17:36:44 tid=30932 @ svcmgr_fini() [server/src/unifyfs_service_manager.c:303] unlocking service manager state
2020-06-15T17:36:44 tid=30932 @ main() [server/src/unifyfs_server.c:406] cleaning run state
2020-06-15T17:36:44 tid=30932 @ unifyfs_exit() [server/src/unifyfs_server.c:594] finalizing kvstore service
2020-06-15T17:36:44 tid=30932 @ unifyfs_exit() [server/src/unifyfs_server.c:599] stopping rpc service
2020-06-15T17:36:45 tid=30932 @ unifyfs_exit() [server/src/unifyfs_server.c:603] stopping metadata service
2020-06-15T17:36:45 tid=30932 @ remove_mdhim_db_filetree() [server/src/unifyfs_metadata.c:184] remove MDHIM DB filetree at /var/tmp/mdhim
2020-06-15T17:36:45 tid=30932 @ unifyfs_exit() [server/src/unifyfs_server.c:611] all done!
```

**Comment by adammoody** [comment]
The cleanup_application code is here:
https://github.com/LLNL/UnifyFS/blob/0dc1e543861768aade9c2d01732f3d69d417f90a/server/src/unifyfs_server.c#L673
Maybe getting wedged trying to acquire a lock?

**Comment by MichaelBrim** [comment]
I think @adammoody is right. The server shutdown code acquires the mutex protecting application state before calling `cleanup_application()`, but inside that function there ends up being a call to `get_application()` (via `get_app_client()`) that also tries to acquire the mutex. The argobots mutex is not recursive, unlike the previous pthread mutex we were using before. I'll work on a fix.

**Comment by MichaelBrim** [comment]
Fix should be in PR #522

**Comment by CamStan** [comment]
Resolved by #522
